cranelift/codegen/src/machinst/abi.rs

```diff
@@ -1173,19 +1176,21 @@ fn gen_store_stack_multi<M: ABIMachineSpec>(
     ret
 }
 
-pub(crate) fn ensure_struct_return_ptr_is_returned(sig: &ir::Signature) -> ir::Signature {
-    let params_structret = sig
-        .params
-        .iter()
-        .find(|p| p.purpose == ArgumentPurpose::StructReturn);
-    let rets_have_structret = sig.returns.len() > 0
-        && sig
-            .returns
-            .iter()
-            .any(|arg| arg.purpose == ArgumentPurpose::StructReturn);
+/// If the signature needs to be legalized, the return the struct-return
+/// parameter that should be prepended to its returns. Otherwise, return `None`.
+fn missing_struct_return(sig: &ir::Signature) -> Option<ir::AbiParam> {
+    let struct_ret_index = sig.special_param_index(ArgumentPurpose::StructReturn)?;
+    if !(sig.returns.len() > 0 && sig.uses_special_return(ArgumentPurpose::StructReturn)) {
```

Comment:
I haven't looked at the implementation of `uses_special_return`. Why does it need to be guarded to only run when there are some return values?

Maybe the answer is just that the old implementation had this guard. But I think the old implementation, using `Iterator::any`, shouldn't have needed this either: `any` returns `false` on an empty iterator.

Reply:
It isn't that `uses_special_return` forces this, it is the pre-existing logic of when we need to legalize, and I have no idea why it is like that, I just faithfully kept it the same.

Reply:
Yeah, it looks like it was an over-eager guard that wasn't actually necessary. Latest version looks good.